### PR TITLE
feat: Add ability to tag network-interface using Launch Template

### PR DESCRIPTION
**Pull request by nikitacr7**
# PR o'clock

## Description

The PR should resolve: https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1361

Please explain the changes you made here and link to any relevant issues.

### Checklist

- [ ] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by bryantbiggs** [comment]
FYI - we'll need to bump the min AWS provider version to [3.53.0](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md#3530-august-05-2021) for this change and we'll also want to update an example to show/test/validate its use

**Comment by antonbabenko** [comment]
I guess there is no better reviewer than @lisfo4ka who added a similar feature a few days ago :)

**Comment by nikitacr7** [comment]
@bryantbiggs @antonbabenko 

Should I update AWS provider version in my PR or it will be updated by you? I've added example for Managed Node Group launch template. 

Worker group launch template is already have example with tags

**Comment by nikitacr7** [comment]
@antonbabenko What are the next steps?

**Comment by antonbabenko** [comment]
Looks good to me, let's hear if @lisfo4ka agrees (hopefully during the next couple of days she replies).

**Comment by lisfo4ka** [comment]
Hi @nikitacr7 ,
Thanks for delivering the latest features from the AWS provider, it's a really useful pull request.

Some notes - at the moment, we can define tags on the global level via `var.tags` map and more specific via `var.worker_groups_launch_template` list of maps and the `local.workers_group_default`.

As I can see, your changes will add `var.tags` to the ENIs of the worker groups created from the launch template, which is great.
But you exclude all the additional tags from the `var.worker_groups_launch_template` and the `local.workers_group_default`, which looks like copy-pasting from instance `tag_specification`, but doesn't make sense here. Worker group instances inherit additional tags from the ASG with `propagate_at_launch=true `attribute. But unfortunately, there is no option to inherit ENI's tags from the ASG in the same way, so we'll prefer to do that by merging all the tags for ENI in the `tag_specifications`.

You can refer to the following PR [#450](https://github.com/terraform-aws-modules/terraform-aws-eks/pull/450) if you'd like to tag ENIs with `var.tags` only.
Or take into account this one #1397 to add more specific tags as well.

For the `node_groups` everything looks fine except indents, please fix it as well.

Also, I believe it's pretty important to include AWS provider version pinning exactly in the same commit with this change.

Please, fill free to ask any questions to get more details or let me know if I've misunderstood something. 

FYI @antonbabenko 

BR,
Olesia

File: modules/node_groups/launch_template.tf [hunk 1]
```diff
@@ -103,6 +103,20 @@ resource "aws_launch_template" "workers" {
     )
   }
 
+  # Supplying custom tags to EKS instances ENI's is another use-case for LaunchTemplates
+  tag_specifications {
+    resource_type = "network-interface"
+
+    tags = merge(
+    var.tags,
```
Comment (lisfo4ka) [inline modules/node_groups/launch_template.tf:111]: please, add indents

File: workers_launch_template.tf [hunk 2]
```diff
@@ -573,6 +573,24 @@ resource "aws_launch_template" "workers_launch_template" {
...
+
+    tags = merge(
+      {
+        "Name" = "${coalescelist(aws_eks_cluster.this[*].name, [""])[0]}-${lookup(
+          var.worker_groups_launch_template[count.index],
+          "name",
+          count.index,
+        )}-eks_asg"
+      },
+      { for tag_key, tag_value in var.tags :
```
Comment (lisfo4ka) [inline workers_launch_template.tf:587]: looks like doesn't make sense here

File: workers_launch_template.tf [hunk 3]
```diff
@@ -573,6 +573,24 @@ resource "aws_launch_template" "workers_launch_template" {
...
+        tag_key => tag_value
+        if tag_key != "Name" && !contains([for tag in lookup(var.worker_groups_launch_template[count.index], "tags", local.workers_group_defaults["tags"]) : tag["key"]], tag_key)
```
Comment (lisfo4ka) [inline workers_launch_template.tf:589]: looks like doesn't make sense here

File: modules/node_groups/launch_template.tf (hunk 1 above)
Comment (nikitacr7) [reply modules/node_groups/launch_template.tf]: @lisfo4ka I've added indents. Thanks

File: workers_launch_template.tf (hunk 3 above)
Comment (nikitacr7) [reply workers_launch_template.tf]: Code has been updated

**Comment by nikitacr7** [comment]
Hi @lisfo4ka @antonbabenko 
Thanks for reviewing my PR. I really appreciate it.

 I've updated Terraform code to address all comments. Could you please review it once again? 

Btw, AWS provider version has been already updated by @antonbabenko in master branch. 

Best regards,
Nikita.

**Review by lisfo4ka — APPROVED** [review]
Hi @nikitacr7,

Thanks for your contribution. Now it looks fine to me. I've tested it with terraform apply and everything works properly.

@antonbabenko, approve from my side)

**Comment by antonbabenko** [comment]
Thanks @nikitacr7 and @lisfo4ka for the review!

[v17.12.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/v17.12.0) has been just released.

**Comment by nikitacr7** [comment]
@antonbabenko @lisfo4ka 

I just want to let you know that changes implemented in the PR don't work due to internal issue at AWS side. AWS support has created issue in container roadmap: https://github.com/aws/containers-roadmap/issues/1496

**Pull request merged**